# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/gamepad/GamepadPlatformService.cpp
# Commit: 2d192d43277b
# Full Hash: 2d192d43277bb6231b9145572d0059c5964f9d96
# Author: Chris Martin <cmartin@mozilla.com>
# Date: 2020-12-15 09:29:54
# Description:
#   Bug 1681750 - Fix random GamepadPlatformService::RemoveChannelParent crashes r=haik
#   
#   The most likely cause for the crashes seems like IPDL intermittently calling GamepadEventChannelParent::ActorDestroy() multiple times on the same object.
#   
#   This adds a wrapper to stop that behavior (if it's the root cause), and also to fire off release assertions at several other potential causes.
# ==============================================================================

diff -r 6de49379342e -r 2d192d43277b dom/gamepad/GamepadPlatformService.cpp
--- a/dom/gamepad/GamepadPlatformService.cpp	Mon Dec 14 15:13:35 2020 +0000
+++ b/dom/gamepad/GamepadPlatformService.cpp	Tue Dec 15 00:01:19 2020 +0000
@@ -264,7 +264,7 @@
     const RefPtr<GamepadEventChannelParent>& aParent) {
   MutexAutoLock autoLock(mMutex);
 
-  MOZ_ASSERT(!mChannelParents.Contains(aParent));
+  MOZ_RELEASE_ASSERT(!mChannelParents.Contains(aParent));
   mChannelParents.AppendElement(aParent);
 
   // Inform the new channel of all the gamepads that have already been added
@@ -279,7 +279,7 @@
     GamepadEventChannelParent* aParent) {
   MutexAutoLock autoLock(mMutex);
 
-  MOZ_ASSERT(mChannelParents.Contains(aParent));
+  MOZ_RELEASE_ASSERT(mChannelParents.Contains(aParent));
 
   // If there is only one channel left, we destroy the singleton instead of
   // unregistering the channel
@@ -318,7 +318,8 @@
   // is created or removed in Background thread
   AssertIsOnBackgroundThread();
   MOZ_ASSERT(aParent);
-  MOZ_ASSERT(gGamepadPlatformServiceSingleton);
+
+  MOZ_RELEASE_ASSERT(gGamepadPlatformServiceSingleton);
 
   // RemoveChannelParentInternal will refuse to remove the last channel
   // In that case, we should destroy the singleton
@@ -328,9 +329,14 @@
 
   GamepadMonitoringState::GetSingleton().Set(false);
   StopGamepadMonitoring();
+  // At this point, any monitor threads should be stopped so we don't need
+  // synchronization
 
-  // At this point, any monitor threads should be stopped and the only
-  // reference to the singleton should be the global one
+  // We should never be destroying the singleton with event channels left in it
+  MOZ_RELEASE_ASSERT(
+      gGamepadPlatformServiceSingleton->mChannelParents.Length() == 1);
+
+  // The only reference to the singleton should be the global one
   MOZ_RELEASE_ASSERT(gGamepadPlatformServiceSingleton->mRefCnt.get() == 1);
 
   gGamepadPlatformServiceSingleton = nullptr;